### PR TITLE
lease: guard 'Lease.itemSet' from concurrent writes

### PR DESCRIPTION
**Pull request by gyuho**
Fix https://github.com/coreos/etcd/issues/7448.

Affected if etcd builds with Go 1.8+.

### CODE REVIEW

File: lease/lessor.go [hunk 1]
```diff
@@ -515,6 +520,9 @@ func (l *Lease) expired() bool {
 }
 
 func (l *Lease) persistTo(b backend.Backend) {
+	l.mu.RLock()
```
Comment (heyitsanthony) [inline lease/lessor.go]: doesn't need to be acquired since not touching itemSet?

File: lease/lessor.go [hunk 2]
```diff
@@ -530,12 +538,15 @@ func (l *Lease) persistTo(b backend.Backend) {
 
 // TTL returns the TTL of the Lease.
 func (l *Lease) TTL() int64 {
-	return l.ttl
+	l.mu.RLock()
```
Comment (heyitsanthony) [inline lease/lessor.go]: doesn't need to be acquired since not touching itemSet?

File: lease/lessor.go [hunk 3]
```diff
@@ -501,6 +505,7 @@ func (le *lessor) initAndRecover() {
 }
 
 type Lease struct {
+	mu  sync.RWMutex
```
Comment (heyitsanthony) [inline lease/lessor.go]: `// mu protects concurrent accesses to itemSet`?

File: lease/lessor_test.go [hunk 4]
```diff
@@ -77,6 +79,52 @@ func TestLessorGrant(t *testing.T) {
...
+	wg.Add(itemn)
+	for i := 0; i < itemn; i++ {
+		go func() {
+			defer wg.Done()
+			l.Keys()
+		}()
+	}
+
+	<-donec
+}
```
Comment (heyitsanthony) [inline lease/lessor_test.go]: `wg.Wait()`?

File: lease/lessor.go [hunk 5]
```diff
@@ -530,12 +538,15 @@ func (l *Lease) persistTo(b backend.Backend) {
...
+	l.mu.RLock()
+	ttl := l.ttl
+	l.mu.RUnlock()
+	return ttl
 }
 
 // refresh refreshes the expiry of the lease.
 func (l *Lease) refresh(extend time.Duration) {
-	t := monotime.Now().Add(extend + time.Duration(l.ttl)*time.Second)
+	t := monotime.Now().Add(extend + time.Duration(l.TTL())*time.Second)
```
Comment (heyitsanthony) [inline lease/lessor.go]: can stay the same?

**Comment by gyuho** [comment]
@heyitsanthony Yeah fixed it to only protect `itemSet`. PTAL. Thanks!

**Comment by heyitsanthony** [comment]
lgtm. Thanks!

**Comment by fanminshi** [comment]
change this 
```
func (le *lessor) Revoke(id LeaseID) error {
...
// not under le.mu at here.
// so detach() can operate on l.itemset at the same time revoke() loops l.itemSet. 
keys := make([]string, 0, len(l.itemSet))
	for item := range l.itemSet {
		keys = append(keys, item.Key)
	}
...
```
https://github.com/gyuho/etcd/blob/508c83becef22fc4d08e8c96039d5ac06ded749f/lease/lessor.go#L255-L258

to
```
keys := l.Keys()
```

**Comment by fanminshi** [comment]
lgtm thanks!

**Pull request merged**